#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: Fix for PR target/26743 (hppa)
# DP: Backport from gcc trunk 2006-04-12 by tausq@debian.org

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

From: "John David Anglin" <dave@hiauly1.hia.nrc.ca>
Sender: gcc-patches-owner@gcc.gnu.org
To: gcc-patches@gcc.gnu.org
Subject: [committed] Define symbolic constants for maximum branch offsets in pa.md
Date: Tue, 11 Apr 2006 14:34:18 -0400 (EDT)

The enclosed change is to facilitate changing the maximum branch
offsets.  There should be no change in functionality.

Tested on hppa-unknown-linux-gnu.  Applied to trunk.

Dave

2006-04-11  John David Anglin  <dava.anglin@nrc-cnrc.gc.ca>

	* pa.md (MAX_12BIT_OFFSET, MAX_17BIT_OFFSET): Define and use new
	symbolic constants in branch patterns.

--- gcc/config/pa/pa.md.orig	2006-04-19 11:33:06.396048656 +0200
+++ gcc/config/pa/pa.md	2006-04-19 11:35:52.790752816 +0200
@@ -1,6 +1,6 @@
 ;;- Machine description for HP PA-RISC architecture for GCC compiler
 ;;   Copyright (C) 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001,
-;;   2002, 2003, 2004, 2005 Free Software Foundation, Inc.
+;;   2002, 2003, 2004, 2005, 2006 Free Software Foundation, Inc.
 ;;   Contributed by the Center for Software Science at the University
 ;;   of Utah.
 
@@ -52,6 +52,16 @@
    (UNSPECV_LONGJMP	5)	; builtin_longjmp
   ])
 
+;; Maximum pc-relative branch offsets.
+
+;; These numbers are a bit smaller than the maximum allowable offsets
+;; so that a few instructions may be inserted before the actual branch.
+
+(define_constants
+  [(MAX_12BIT_OFFSET     8184)	; 12-bit branch
+   (MAX_17BIT_OFFSET   262100)	; 17-bit branch
+  ])
+
 ;; Insn type.  Used to default other attribute values.
 
 ;; type "unary" insns have one input operand (1) and one output operand (0)
@@ -1713,10 +1723,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1742,10 +1752,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1769,10 +1779,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1798,10 +1808,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1824,10 +1834,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1853,10 +1863,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -1882,10 +1892,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -1910,10 +1920,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -1938,10 +1948,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -1966,10 +1976,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -1994,10 +2004,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2022,10 +2032,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2050,10 +2060,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2078,10 +2088,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2107,10 +2117,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2135,10 +2145,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2163,10 +2173,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2191,10 +2201,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2219,10 +2229,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2247,10 +2257,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2275,10 +2285,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2303,10 +2313,10 @@
 [(set_attr "type" "cbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -2355,7 +2365,7 @@
 [(set_attr "type" "fbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 32)
@@ -2399,7 +2409,7 @@
 [(set_attr "type" "fbranch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 12)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 28)
@@ -7142,11 +7152,11 @@
     (cond [(eq (symbol_ref "jump_in_call_delay (insn)") (const_int 1))
 	   (if_then_else (lt (abs (minus (match_dup 0)
 					 (plus (pc) (const_int 8))))
-			     (const_int 8184))
+			     (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (const_int 8))
 	   (lt (abs (minus (match_dup 0) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 4)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 20)
@@ -8968,10 +8978,10 @@
 ;; Short branch has length of 4
 ;; Long branch has length of 8, 20, 24 or 28
 	(cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -8984,10 +8994,10 @@
 	(if_then_else (eq_attr "alternative" "1")
 	  (if_then_else (lt (match_dup 3) (pc))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 24))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 24)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 24))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 28)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 44)
@@ -8995,10 +9005,10 @@
 		    (const_int 40)]
 		  (const_int 48))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 24)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 28)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 44)
@@ -9010,10 +9020,10 @@
 ;; Extra goo to deal with additional reload insns.
 	(if_then_else (lt (match_dup 3) (pc))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9021,10 +9031,10 @@
 		    (const_int 28)]
 		  (const_int 36))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9051,10 +9061,10 @@
 ;; Short branch has length of 4
 ;; Long branch has length of 8, 20, 24 or 28
         (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9067,10 +9077,10 @@
 	(if_then_else (eq_attr "alternative" "1")
 	  (if_then_else (lt (match_dup 3) (pc))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9078,10 +9088,10 @@
 		    (const_int 28)]
 		  (const_int 36))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9092,10 +9102,10 @@
 ;; Loop counter in memory or sar case.
 ;; Extra goo to deal with additional reload insns.
 	(cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		   (const_int 8184))
+		   (const_int MAX_12BIT_OFFSET))
 		(const_int 8)
 		(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		  (const_int 262100))
+		  (const_int MAX_17BIT_OFFSET))
 		(const_int 12)
 		(ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		(const_int 28)
@@ -9123,10 +9133,10 @@
 ;; Short branch has length of 4
 ;; Long branch has length of 8
         (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9139,10 +9149,10 @@
 	(if_then_else (eq_attr "alternative" "1")
 	  (if_then_else (lt (match_dup 3) (pc))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 12))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9150,10 +9160,10 @@
 		    (const_int 28)]
 		  (const_int 36))
 	     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 8184))
+		      (const_int MAX_12BIT_OFFSET))
 		    (const_int 12)
 		    (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		      (const_int 262100))
+		      (const_int MAX_17BIT_OFFSET))
 		    (const_int 16)
 		    (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		    (const_int 32)
@@ -9164,10 +9174,10 @@
 ;; Loop counter in memory or SAR case.
 ;; Extra goo to deal with additional reload insns.
 	(cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		   (const_int 8184))
+		   (const_int MAX_12BIT_OFFSET))
 		(const_int 8)
 		(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-		  (const_int 262100))
+		  (const_int MAX_17BIT_OFFSET))
 		(const_int 12)
 		(ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 		(const_int 28)
@@ -9188,10 +9198,10 @@
 [(set_attr "type" "parallel_branch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 3) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9211,10 +9221,10 @@
 [(set_attr "type" "parallel_branch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9234,10 +9244,10 @@
 [(set_attr "type" "parallel_branch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9257,10 +9267,10 @@
 [(set_attr "type" "parallel_branch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
@@ -9280,10 +9290,10 @@
 [(set_attr "type" "parallel_branch")
  (set (attr "length")
     (cond [(lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 8184))
+	       (const_int MAX_12BIT_OFFSET))
 	   (const_int 4)
 	   (lt (abs (minus (match_dup 2) (plus (pc) (const_int 8))))
-	       (const_int 262100))
+	       (const_int MAX_17BIT_OFFSET))
 	   (const_int 8)
 	   (ne (symbol_ref "TARGET_PORTABLE_RUNTIME") (const_int 0))
 	   (const_int 24)
